### ISSUE TITLE
Possible to define hotkey on dead key (like grave accent)?

### ISSUE THREAD

**Issue by elmart** [Hammerspoon/hammerspoon#1037]
I suspect when hammerspoon process events, it's too late for that, but just want to check.

**Comment by asmagill** [comment]
Pardon my inexperience with internationalization and non-US keyboards -- it's something I've only recently begun to work with myself, so...

Does the "grave accent" correspond to an actual key on the keyboard?  Or is it applied as part of a key combination or by holding a key down for a certain amount of time?

Hotkey is limited to only those key's which correspond to actual physical keys on the keyboard; and if you're needing to use different keyboard layouts regularly, it's probably safest to refer to the key position by its number rather than it's name (e.g. for the "A" character on a US keyboard layout, the keycode number is 0) so I could apply a hotkey with `hs.hotkey.bind({"cmd"}, 0, ...)` rather than `hs.hotkey.bind({"cmd"}, "a", ...)` and it would always be the physical key corresponding to the first position on the second row of letters no matter what layout I currently have active.

You can get a list of the currently identified "physical" keys by typing `hs.inspect(hs.keycodes.map)` into the Hammerspoon console. You can also look at `/System/Library/Frameworks/Carbon.framework/Frameworks/HIToolbox.framework/Headers/Events.h` if you have the XCode command line tools installed and look for the section on Virtual keycodes (around line 200)... this can be useful for identifying numbers for keys which are being missed by `hs.keycodes.map` or correspond to non-traditional keys that haven't been added to the map yet -- we've had a couple in the past and I wouldn't be surprised if there may be a few others that are on some keyboards, but not "Standard US" or "Standard 104 Key" keyboards, against which the earliest versions of these modules were written and tested.

If the "dead keys" you refer to come from combinations or from holding a key down for an certain length of time, then `hs.eventtap` will be the only way to detect them.

**Comment by elmart** [comment]
I use a Spanish keyboard layout, like the one here: http://www.apple.com/es/shop/product/MLA22Y/A/magic-keyboard-español?afid=p238%7CsyaBLcH5G-dt_mtid_187079nc38483_pcrid_95121011584_&cid=aos-es-kwgo-pla-btb-slid--product-MLA22Y/A-ES

Grave accent is the key right next to P. You also call that backtick, or backquote, in English.
In an English layout, it's in the upper left corner, if I'm not mistaken. 

Regarding dead keys, please see here: https://en.m.wikipedia.org/wiki/Dead_key
If your layout is also international US, then backtick is also dead for you. 

So, given that a dead key doesn't generate a character until the next key is pressed (and this is presumably handled by the keyboard driver, before passing the character to the focused application), I'm not sure if Hammerspoon's hotkey mechanism has even a chance to bind there.

Please tell me if I didn't make my point clear.

**Comment by asmagill** [comment]
I'm going to guess that the layout on my laptop is not the international US one... I've heard reference to such before, but I don't know the difference between it and the standard US one.  However I can tell you that the backtick is not a dead key on my keyboard/layout... although I can use it like a "dead key" by holding the option key while tapping it and then following it with a letter like `a` or `e` to get `à` or `è`...

Hmm, the key on my keyboard to the right of the 'P' is a left bracket or "[", and `hs.keycodes.map["["]` returns 33, so... see what this does for you:

~~~lua
j = hs.hotkey.bind({"cmd", "shift"}, 33, function() print("hello!") end, function() print("goodbye!") end)
~~~

What happens when you hold the cmd and shift key modifiers and then tap and then release the key to the right of the "P"?

**Comment by elmart** [comment]
There's something weird going on here.
Inspecting `hs.keycodes.map` I see backtick has code 39 on my system.
I mean I get ```hs.keycodes.map['`'] == 39``` and ```hs.keycodes.map[39] == '`'``` on the console.
But trying `hs.hotkey.bind({}, 39, ...)` does nothing.
However... I also get ```hs.keycodes.map[33] == '`'```...
and `hs.hotkey.bind({}, 33, ...)` does in fact work!

Except one thing: I want to use that key as a modifier, when hold down while pressing another key, and generating normal backtick when pressed alone. For that, I used hs.hotkey.modal, in the following way:
```lua
myModifierMode = hs.hotkey.modal.new()

myModifierMode:bind({}, 'a', 
  function()
    myModifierMode.triggered = true
    -- do something
  end
)

myModifier = hs.hotkey.bind({}, 33,
  function()
    myModifierMode:enter()
    myModifierMode.triggered = false
  end,
  function()
    myModifierMode:exit()
    if not myModifierNode.triggered then
      hs.eventap.keyStroke({}, 33)
    end
  end
)
```
That is, pressing <kbd>`</kbd> + <kbd>a</kbd> should do something.

While pressing and releasing  <kbd>`</kbd> alone should just generate a normal backtick.

Now, first part (use as modifier) works, but second (generating backtick alone) fails (I get an `a` instead).
I have that same arrangement working with other *normal* keys, so I suppose the difference here is that the keystroke generation is misbehaving because of some reason having to do with backtick being a dead key.

Phew ;-)

**Comment by asmagill** [comment]
In combination with the \` key, I'm not sure if it's actually an "a" that's being sent since it's acting as a "dead" key on your keyboard... you'll need to use eventtap to see what's really going on...  something like this:

~~~lua
local counter = 0
local watcher
watcher = hs.eventtap.new({hs.eventtap.event.types.keyDown, hs.eventtap.event.types.keyUp}, function(ev)
    local keycode = ev:getKeyCode()
    local labels = {
        [hs.eventtap.event.types.keyDown] = "down",
        [hs.eventtap.event.types.keyUp] = "up",
    }
    print(labels[ev:getType()], keycode, ev:getCharacters(), hs.keycodes.map[keycode])
    counter = counter + 1
    if counter == 5 then
        watcher:stop()
    end
    return false
end):start()
~~~

Copy and paste this into the Hammerspoon console.  Adjust the `if counter == 5 then` line if you want to see more events... 5 allows for the return key which will be released when you enter this into the console, then the keyDown for "\`", the keyDown for "a", the keyUp for "a", and the keyUp for "\`".

I'm most curious about the keyDown and keyUp lines for the "a" (i.e. not the "\`") character... what keycode and character do they show?

**Comment by elmart** [comment]
When I press and release <kbd>`</kbd>, I get this on the console using your script:
```
down	33		`
up	33		`
18:13:48     hotkey: Disabled hotkey `
             hotkey: Enabled hotkey `
down	0	a	a
up	0	a	a
```

So, it seems my problem is that `hs.eventtap.keyStroke({}, 33)` generates key events with keycode 0, which translates to the 'a' character (as `hs.keycodes.map[0] == 'a'`).

Can something be done about this?

**Comment by elmart** [comment]
BTW, this is my keycodes map:
```
> hs.inspect(hs.keycodes.map)
-- Loading extension: inspect
-- Loading extension: keycodes
{ "s", "d", "f", "h", "g", "z", "x", "c", "v", "º", "b", "q", "w", "e", "r", "y", "t", "1", "2", "3", "4", "6", "5", "¡", "9", "7", "'", "8", "0", "+", "o", "u", "`", "i", "p", "return", "l", "j", "`", "k", "ñ", "ç", ",", "-", "n", "m", ".", "tab", "space", "<", "delete", nil, "escape", nil, nil, nil, nil, nil, nil, nil, nil, nil, nil, "f17", "pad.", nil, "pad*", nil, "pad+", nil, "padclear", nil, nil, nil, "pad/", "padenter", nil, "pad-", "f18", "f19", "pad=", "pad0", "pad1", "pad2", "pad3", "pad4", "pad5", "pad6", "pad7", "f20", "pad8", "pad9", nil, nil, nil, "f5", "f6", "f7", "f3", "f8", "f9",
  [0] = "a",
  [103] = "f11",
  [105] = "f13",
  [106] = "f16",
  [107] = "f14",
  [109] = "f10",
  [111] = "f12",
  [113] = "f15",
  [114] = "help",
  [115] = "home",
  [116] = "pageup",
  [117] = "forwarddelete",
  [118] = "f4",
  [119] = "end",
  [120] = "f2",
  [121] = "pagedown",
  [122] = "f1",
  [123] = "left",
  [124] = "right",
  [125] = "down",
  [126] = "up",
  ["'"] = 27,
  ["+"] = 30,
  [","] = 43,
  ["-"] = 44,
  ["."] = 47,
  ["0"] = 29,
  ["1"] = 18,
  ["2"] = 19,
  ["3"] = 20,
  ["4"] = 21,
  ["5"] = 23,
  ["6"] = 22,
  ["7"] = 26,
  ["8"] = 28,
  ["9"] = 25,
  ["<"] = 50,
  ["`"] = 39,
  a = 0,
  b = 11,
  c = 8,
  d = 2,
  delete = 51,
  down = 125,
  e = 14,
  end = 119,
  escape = 53,
  f = 3,
  f1 = 122,
  f10 = 109,
  f11 = 103,
  f12 = 111,
  f13 = 105,
  f14 = 107,
  f15 = 113,
  f16 = 106,
  f17 = 64,
  f18 = 79,
  f19 = 80,
  f2 = 120,
  f20 = 90,
  f3 = 99,
  f4 = 118,
  f5 = 96,
  f6 = 97,
  f7 = 98,
  f8 = 100,
  f9 = 101,
  forwarddelete = 117,
  g = 5,
  h = 4,
  help = 114,
  home = 115,
  i = 34,
  j = 38,
  k = 40,
  l = 37,
  left = 123,
  m = 46,
  n = 45,
  o = 31,
  p = 35,
  ["pad*"] = 67,
  ["pad+"] = 69,
  ["pad-"] = 78,
  ["pad."] = 65,
  ["pad/"] = 75,
  pad0 = 82,
  pad1 = 83,
  pad2 = 84,
  pad3 = 85,
  pad4 = 86,
  pad5 = 87,
  pad6 = 88,
  pad7 = 89,
  pad8 = 91,
  pad9 = 92,
  ["pad="] = 81,
  padclear = 71,
  padenter = 76,
  pagedown = 121,
  pageup = 116,
  q = 12,
  r = 15,
  return = 36,
  right = 124,
  s = 1,
  space = 49,
  t = 17,
  tab = 48,
  u = 32,
  up = 126,
  v = 9,
  w = 13,
  x = 7,
  y = 16,
  z = 6,
  ["¡"] = 24,
  ["º"] = 10,
  ["ç"] = 42,
  ["ñ"] = 41
}
```

**Comment by asmagill** [comment]
According to the sequence of events you've posted, it looks like you pushed \` down, then released it, *then* pushed `a` down and then released it... if this is the case, then your original code did exactly what it was supposed to do:  when \` is down, `a` is defined (bound modal keys are enabled), but when \` is released, `a` is no longer defined (bound modal keys are disabled).

Personally I've never used modal keys in this fashion (with a key-release function that exits the modal state), but then I've never tried to match multiple keys being held down at the same time.

However, if you are certain that you didn't release the \` key until *after* you pressed and released the `a` key, then it appears that the OS is overriding the events and changing their order during its facilitation of the dead-key behavior.

Personally I've never used modal keys in this fashion (with a key-release function that exits the modal state), but then I've never tried to match multiple keys being held down at the same time... I would more likely do something like this:

~~~lua
-- note I can't test this right now, so it may need tweaking, but I think it should be close...

myModifierMode = hs.hotkey.modal.new({}, 33)

function myModifierMode:entered()
    -- if the user holds the key for more than a second before tapping another key, auto-exit
    -- the modal state
    myModifierMode.timedRelease = hs.timer.doAfter(1, function() myModifierMode:exit() end
    myModifierMode.triggered = false
end

function myModifierMode:exited()
    if not myModifierMode.triggered then
       myModifier:disable()
       hs.eventtap.keyStroke({}, 33)
       myModifier:enable()
    end
    myModifierMode.timedRelease = nil
    myModifierMode.triggered = nil
end

myModifierMode:bind({}, 'a', function()
    myModifierMode.timedRelease:stop()
    myModifierMode.triggered = true

    -- do something

    myModifierMode:exit()
end)
~~~

**Comment by elmart** [comment]
> According to the sequence of events you've posted, it looks like you pushed ` down, then released it, then pushed a down and then released it... if this is the case, then your original code did exactly what it was supposed to do: when ` is down, a is defined (bound modal keys are enabled), but when ` is released, a is no longer defined (bound modal keys are disabled).

I think you're getting confused, focusing on the first case I mentioned (pressing <kbd>`</kbd> + <kbd>a</kbd>, then releasing both). That case works right for me.

What is not working is the second case (pressing and releasing <kbd>`</kbd) alone). In that case, I want to do what that key would do if no hammerspoon was involved. This is, generating a normal bactick key press.

I'm managing that case in the second function of:
```
myModifier = hs.hotkey.bind({}, 33,
  function()
    myModifierMode:enter()
    myModifierMode.triggered = false
  end,
  function()
    myModifierMode:exit()
    if not myModifierMode.triggered then
      myModifier:disable()
      hs.eventtap.keyStroke({}, 33)
      myModifier:enable()
    end
  end
)
```

So, it's `hs.eventap.keyStroke({}, 33)` which is executing when I release backtick.
And it's that that is generating an 'a'.

In fact, if I do `hs.eventap.keyStroke({}, 33)` directly into the console, I get an 'a' right there (in the following console line).

In other words, it was an unfortunate coincidence that I chose the example <kbd>`</kbd> + <kbd>a</kbd> for the first case, and that my problem involves another 'a'. But they have nothing to do one another. There are no problems with my code. It's executing as intended.
The problem seems to be that, for some reason, `hs.eventap.keyStroke({}, 33)` is generating key events with keycode 0, instead of 33. And then hammerspoon translates that 0 to an 'a' (as hs.keycodes.map[0] is 'a' in my system).

Sorry for all the confusion. Do you get my point now?

**Comment by asmagill** [comment]
Well, unfortunately I just discovered that `hs.eventtap.keyStroke` doesn't handle numbers from `hs.keycodes.map` correctly... `hs.hotkey` does, but eventtap doesn't... that's an oversight and I'll be creating an issue so that it can be fixed in the next release... but in the meantime, see if this works for you:

Add the following to your `init.lua` file:

~~~lua
hs.eventtap.keyStroke = function(mods, num)
    if type(num) == "number" then
        -- "a" is filler -- it is replaced by the value in the setKeyCode method
        hs.eventtap.event.newKeyEvent(mods, "a", true):setKeyCode(math.tointeger(num)):post()
        hs.eventtap.event.newKeyEvent(mods, "a", false):setKeyCode(math.tointeger(num)):post()
    else
        hs.eventtap.event.newKeyEvent(mods, string.lower(num), true):post()
        hs.eventtap.event.newKeyEvent(mods, string.lower(num), false):post()
    end
end
~~~

Now `hs.eventtap.keyStroke` should work as you expect it to when numbers are specified.

**Comment by elmart** [comment]
That worked perfectly! 👍 
Thanks very much for your patience with this.
Closing.
